src/SAWScript/Crucible/Common/Setup/Value.hs

```diff
@@ -0,0 +1,211 @@
+--   no backend-specific information, then the corresponding extension field
+--   should have a @type instance@ that evaluates to @()@.
+--
+-- - If a SetupValue constructor is not used in a particular backend, its
+--   corresponding extension field should have a @type instance@ that evaluates
+--   to Void. Any code that pattern matches on the constructor should then
+--   dispatch on the Void value using the 'absurd' function.
+
+type family XSetupNull ext
+type family XSetupStruct ext
+type family XSetupArray ext
+type family XSetupElem ext
+type family XSetupField ext
+type family XSetupGlobal ext
+type family XSetupCast ext
+type family XSetupUnion ext
+type family XSetupGlobalInitializer ext
+
+-- | From the manual: \"The SetupValue type corresponds to values that can occur
+-- during symbolic execution, which includes both 'Term' values, pointers, and
+-- composite types consisting of either of these (both structures and arrays).\"
+data SetupValue ext where
+  SetupVar    :: AllocIndex -> SetupValue ext
+  SetupTerm   :: TypedTerm -> SetupValue ext
+  SetupNull   :: XSetupNull ext -> SetupValue ext
+  SetupStruct :: XSetupStruct ext -> [SetupValue ext] -> SetupValue ext
+  SetupArray  :: XSetupArray ext -> [SetupValue ext] -> SetupValue ext
+  SetupElem   :: XSetupElem ext -> SetupValue ext -> Int -> SetupValue ext
+  SetupField  :: XSetupField ext -> SetupValue ext -> String -> SetupValue ext
+  SetupCast   :: XSetupCast ext -> SetupValue ext -> SetupValue ext
+  SetupUnion  :: XSetupUnion ext -> SetupValue ext -> String -> SetupValue ext
+
+  -- | A pointer to a global variable
+  SetupGlobal :: XSetupGlobal ext -> String -> SetupValue ext
+  -- | This represents the value of a global's initializer.
+  SetupGlobalInitializer ::
+    XSetupGlobalInitializer ext -> String -> SetupValue ext
+
+-- | This constraint can be solved for any ext so long as '()', 'Void', and any
+--   other types used in the right-hand sides of extension field
+--   @type instance@s have the constraint. Unfortunately, GHC can't (yet?)
+--   reason over the equations in our closed type family and realize that, so we
+--   must explicitly abstract this reasoning out into a type synonym.
+type SetupValueHas (c :: Type -> Constraint) ext =
+  ( c (XSetupNull ext)
+  , c (XSetupStruct ext)
+  , c (XSetupArray ext)
+  , c (XSetupElem ext)
+  , c (XSetupField ext)
+  , c (XSetupCast ext)
+  , c (XSetupUnion ext)
+  , c (XSetupGlobal ext)
+  , c (XSetupGlobalInitializer ext)
+  )
+
+deriving instance (SetupValueHas Show ext) => Show (SetupValue ext)
+
+-- TypedTerm is neither Eq nor Ord
+-- deriving instance (SetupValueHas Eq ext) => Eq (SetupValue ext)
+-- deriving instance (SetupValueHas Ord ext) => Ord (SetupValue ext)
+
+--------------------------------------------------------------------------------
+-- ** Ghost state
+
+-- TODO: This is language-independent, it should be always-true rather than a
+-- toggle.
+
+-- TODO: documentation
```

Comment:
Are these TODOs supposed to be addressed as part of this PR? It's OK if not, just want to make sure you didn't miss something.

Reply:
These TODOs were copied over when I migrated the code to its new home in this module. That being said, I agree that it's somewhat distasteful to leave these TODOs here. I've opened #1929 to track supporting ghost state in other language backends, and I've replaced the TODOs with a small bit of Haddocks that point to #1929.